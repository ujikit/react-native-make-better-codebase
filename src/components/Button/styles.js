//package import
import { StyleSheet } from 'react-native';

//local import
import { COLORS } from '../../configs';
import { FONT_HEADLINE_H6 } from '../../configs/fonts';

//variable valu
const container = {
  alignItems: 'center',
  borderRadius: 4,
  justifyContent: 'center',
  padding: 16,
};
const text = {
  ...FONT_HEADLINE_H6,
  lineHeight: 18,
};

export default StyleSheet.create({
  containerFilled: {
    ...container,
    backgroundColor: COLORS.primaryOrange,
  },
  containerGhost: {
    ...container,
    backgroundColor: COLORS.primaryWhite,
    borderWidth: 1,
    borderColor: COLORS.primaryOrange,
  },
  containerNude: {
    ...container,
    backgroundColor: COLORS.transparent,
  },
  containerDisableFilled: {
    ...container,
    backgroundColor: COLORS.black50,
  },
  containerDisableGhost: {
    ...container,
    backgroundColor: COLORS.black50,
    borderWidth: 1,
    borderColor: COLORS.black80,
  },
  containerDisableNude: {
    ...container,
    backgroundColor: COLORS.transparent,
  },
  textFilled: {
    ...text,
    color: COLORS.primaryWhite,
  },
  textGhost: {
    ...text,
    color: COLORS.primaryOrange,
  },
  textNude: {
    ...text,
    color: COLORS.primaryOrange,
  },
  textDisableFilled: {
    ...text,
    color: COLORS.black80,
  },
  textDisableGhost: {
    ...text,
    color: COLORS.black60,
  },
  textDisableNude: {
    ...text,
    color: COLORS.black60,
  },
  wrapContainer: {
    overflow: 'hidden',
    width: null,
    borderRadius: 4,
  },
});
