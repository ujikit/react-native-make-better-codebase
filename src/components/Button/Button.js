//package import
import React, { memo } from 'react';
import {
  TouchableOpacity,
  TouchableNativeFeedback,
  Text,
  Platform,
  View,
  ActivityIndicator,
} from 'react-native';
import PropTypes from 'prop-types';

//local import
import styles from './styles';
import { COLORS, STYLES } from '../../configs';

const Button = ({
  children,
  color,
  styleContainer,
  styleText,
  styleWrap,
  disabled,
  disabledColor,
  isLoading,
  isUpperCase,
  loadingTitle,
  onPress,
  title,
  types,
  width,
  height,
  ...props
}) => {
  //variable value
  let containerStyle = styles.containerFilled;
  let textStyle = styles.textFilled;
  let loadingColors = COLORS.primaryWhite;
  let widthWrap = null;
  let widthContainer = null;
  let heightWrap = null;
  let heightContainer = null;

  const colors = disabled ? disabledColor : color;

  if (disabled) {
    switch (types) {
      case 'filled':
        containerStyle = [
          styles.containerDisableFilled,
          color && { backgroundColor: colors },
        ];
        textStyle = styles.textDisableFilled;
        loadingColors = COLORS.black80;
        break;
      case 'ghost':
        containerStyle = [
          styles.containerDisableGhost,
          color && { backgroundColor: colors },
        ];
        textStyle = styles.textDisableGhost;
        loadingColors = COLORS.black60;
        break;
      case 'nude':
        containerStyle = styles.containerDisableNude;
        textStyle = styles.textDisableNude;
        loadingColors = COLORS.black60;
        break;
    }
  } else {
    switch (types) {
      case 'filled':
        containerStyle = [
          styles.containerFilled,
          color && { backgroundColor: colors },
        ];
        textStyle = styles.textFilled;
        loadingColors = COLORS.primaryWhite;
        break;
      case 'ghost':
        containerStyle = [
          styles.containerGhost,
          color && { borderColor: colors },
        ];
        textStyle = [styles.textGhost, color && { color: colors }];
        loadingColors = COLORS.primaryOrange;
        break;
      case 'nude':
        containerStyle = styles.containerNude;
        textStyle = [styles.textNude, color && { color: colors }];
        loadingColors = COLORS.primaryOrange;
        break;
    }
  }

  if (isUpperCase) {
    title = title.toUpperCase();
  }

  if (width !== null && width !== undefined) {
    if (width.toString().match(/%/g)) {
      widthWrap = width;
      widthContainer = '100%';
    } else {
      widthWrap = width;
      widthContainer = width;
    }
  }

  if (height !== null && height !== undefined) {
    if (height.toString().match(/%/g)) {
      heightWrap = height;
      heightContainer = '100%';
    } else {
      heightWrap = height;
      heightContainer = height;
    }
  }

  if (Platform.OS === 'ios') {
    return (
      <View
        style={[
          styles.wrapContainer,
          { width: widthWrap, height: heightWrap },
          styleWrap,
        ]}
      >
        <TouchableOpacity
          style={[
            containerStyle,
            { width: widthContainer, height: heightContainer },
            styleContainer,
          ]}
          disabled={disabled}
          onPress={onPress}
          activeOpacity={0.8}
          {...props}
        >
          {isLoading ? (
            <View style={STYLES.rowAlgCenter}>
              <ActivityIndicator color={loadingColors} />
              {loadingTitle ? (
                <Text style={[textStyle, styleText, STYLES.mrl8]}>
                  {loadingTitle}
                </Text>
              ) : null}
            </View>
          ) : (
            <>
              {children}
              {title ? (
                <Text style={[textStyle, styleText]}>{title}</Text>
              ) : null}
            </>
          )}
        </TouchableOpacity>
      </View>
    );
  } else {
    return (
      <View
        style={[
          styles.wrapContainer,
          { width: widthWrap, height: heightWrap },
          styleWrap,
        ]}
      >
        <TouchableNativeFeedback
          disabled={disabled}
          onPress={onPress}
          useForeground={true}
          background={TouchableNativeFeedback.Ripple('rgba(1, 0, 0, 0.2)')}
          {...props}
        >
          <View
            style={[
              containerStyle,
              { width: widthContainer, height: heightContainer },
              styleContainer,
            ]}
          >
            {isLoading ? (
              <View style={STYLES.rowAlgCenter}>
                <ActivityIndicator color={loadingColors} />
                {loadingTitle ? (
                  <Text style={[textStyle, styleText, STYLES.mrl8]}>
                    {loadingTitle}
                  </Text>
                ) : null}
              </View>
            ) : (
              <>
                {children}
                {title ? (
                  <Text style={[textStyle, styleText]}>{title}</Text>
                ) : null}
              </>
            )}
          </View>
        </TouchableNativeFeedback>
      </View>
    );
  }
};

Button.propTypes = {
  children: PropTypes.node,
  color: PropTypes.string,
  styleContainer: PropTypes.oneOfType([PropTypes.object, PropTypes.any]),
  styleText: PropTypes.oneOfType([PropTypes.object, PropTypes.any]),
  styleWrap: PropTypes.oneOfType([PropTypes.object, PropTypes.any]),
  disabled: PropTypes.bool,
  disabledColor: PropTypes.string,
  isLoading: PropTypes.bool,
  isUpperCase: PropTypes.bool,
  loadingTitle: PropTypes.string,
  onPress: PropTypes.func,
  title: PropTypes.string,
  types: PropTypes.oneOf(['filled', 'ghost', 'nude']),
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

Button.defaultProps = {
  children: <View />,
  color: '',
  styleContainer: {},
  styleText: {},
  styleWrap: {},
  disabled: false,
  disabledColor: COLORS.black60,
  isLoading: false,
  isUpperCase: false,
  loadingTitle: '',
  onPress: () => null,
  title: '',
  types: 'filled',
  width: null,
  height: null,
};

export default memo(Button);
