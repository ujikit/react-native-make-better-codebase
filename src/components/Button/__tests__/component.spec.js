import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Button from '..';

describe('Some component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
