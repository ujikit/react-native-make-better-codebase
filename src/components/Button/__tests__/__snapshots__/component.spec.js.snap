// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Some component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 4,
        "overflow": "hidden",
        "width": null,
      },
      Object {},
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#2D9CDB",
        "borderRadius": 4,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 16,
      }
    }
  >
    <View />
  </View>
</View>
`;
